File: src/main/scala/io/koju/autopos/trade/sale/api/SaleApi.scala
package io.koju.autopos.trade.sale.api

import java.net.URI

import com.codahale.metrics.annotation.Timed
import com.fasterxml.jackson.annotation.JsonView
import io.koju.autopos.catalog.domain.SaleStatus
import io.koju.autopos.kernel.json.Views
import io.koju.autopos.trade.sale.domain.Sale
import io.koju.autopos.trade.sale.repo.SaleLineRepo
import io.koju.autopos.trade.sale.service.SaleService
import io.koju.autopos.web.rest.util.{HeaderUtil, PaginationUtil}
import org.springframework.data.domain.Pageable
import org.springframework.http.{HttpStatus, ResponseEntity}
import org.springframework.web.bind.annotation._


@RestController
@RequestMapping(Array("/api/sales"))
class SaleApi(private val saleService: SaleService,
              private val saleLineRepo: SaleLineRepo) {

  @GetMapping
  @Timed
  @JsonView(Array(classOf[Views.Summary]))
  def getAllSales(@RequestParam("status") status: SaleStatus, pageable: Pageable) = {
    val page = saleService.getSalesWithStatus(status, pageable)
    val headers = PaginationUtil.generatePaginationHttpHeaders(page, "/api/sales")
    new ResponseEntity(page.getContent, headers, HttpStatus.OK)
  }

  @PostMapping
  @Timed
  def createNewSale(): ResponseEntity[Sale] = {
    val newSale = saleService.createNewSale()
    ResponseEntity
      .created(new URI(s"/api/sales/${newSale.getId}"))
      .body(newSale)
  }

  @PutMapping(Array("/{id}"))
  @Timed
  def updateSale(@PathVariable("id") id: Long, @RequestBody sale: Sale): ResponseEntity[Sale] = {
    sale.setId(id)
    val updatedSale = saleService.updateSale(sale)
    ResponseEntity.ok
      .headers(HeaderUtil.createEntityUpdateAlert("sale", updatedSale.getId.toString))
      .body(updatedSale)
  }

  @GetMapping(Array("/{id}"))
  @Timed
  def getSale(@PathVariable("id") saleOp: Sale): ResponseEntity[Sale] = {
    Option(saleOp) match {
      case Some(sale) => ResponseEntity.ok(sale)
      case None => new ResponseEntity[Sale](HttpStatus.NOT_FOUND)
    }
  }


}

File: src/main/kotlin/com/kapilkoju/autopos/trade/sale/api/SaleApi.kt
package com.kapilkoju.autopos.trade.sale.api


import com.fasterxml.jackson.annotation.JsonView
import com.kapilkoju.autopos.kernel.json.Views
import com.kapilkoju.autopos.trade.sale.domain.Sale
import com.kapilkoju.autopos.trade.sale.dto.CreateSaleDto
import com.kapilkoju.autopos.trade.sale.dto.GetSaleDto
import com.kapilkoju.autopos.trade.sale.service.SaleService
import com.kapilkoju.autopos.web.rest.util.HeaderUtil
import com.kapilkoju.autopos.web.rest.util.PaginationUtil
import org.springframework.data.domain.Pageable
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.net.URI

@RestController
@RequestMapping(SaleApi.baseUrl)
class SaleApi(private val saleService: SaleService) {

    @GetMapping

    @JsonView(Views.Summary::class)
    fun getAllSales(getSaleDto: GetSaleDto, pageable: Pageable)
            : ResponseEntity<List<Sale>> {

        val page = saleService.getSalesWithStatus(getSaleDto, pageable)
        val headers = PaginationUtil.generatePaginationHttpHeaders(page, baseUrl)
        return ResponseEntity(page.content, headers, HttpStatus.OK)
    }

    @PostMapping

    fun createNewSale(@RequestBody createSaleDto: CreateSaleDto): ResponseEntity<Sale> {
        val newSale = saleService.createNewSale(createSaleDto)
        return ResponseEntity
                .created(URI("/api/sales/${newSale.getId()}"))
                .body(newSale)
    }

    @PutMapping("/{id}")

    fun updateSale(@PathVariable("id") id: Long, @RequestBody sale: Sale)
            : ResponseEntity<Sale> {
        sale.setId(id)
        val updatedSale = saleService.updateSale(sale)
        return ResponseEntity.ok()
                .headers(HeaderUtil.createEntityUpdateAlert("sale", updatedSale.getId().toString()))
                .body(updatedSale)
    }

    @GetMapping("/{id}")

    fun getSale(@PathVariable("id") sale: Sale): ResponseEntity<Sale> {
        return ResponseEntity.ok(sale)
    }

    @DeleteMapping("/{id}")

    fun deleteSale(@PathVariable("id") sale: Sale)
            : ResponseEntity<Void> {
        saleService.deleteSale(sale)
        val headers = HeaderUtil.createEntityDeletionAlert("sale", sale.getId().toString())
        return ResponseEntity(headers, HttpStatus.OK)
    }

    companion object {
        const val baseUrl = "/api/sales"
    }

}
